### ISSUE TITLE
Consider reverting hierarchical paths for resources

### ISSUE THREAD

**Issue by srosset81** [assemblee-virtuelle/activitypods#220]
In v2, the resources will not have an URL dependant on the container path. It will use the root path, and a UUID slug.

This is allowed by LDP, but goes against Solid specifications: https://github.com/solid/specification/issues/98

What should we do ?

## Advantages of non-hierarchical URIs

- We can move the resource to another container without leaving behind dozens of redirects
- We can put the resources in several containers, and thus benefit from the containers' default WAC permissions
- It's harder to guess what a resource is about

## Advantages of hierarchical URIs

- It's easier for developers to know what the resource is about
- It's easier to find the container, without needing to look for `ldp:contains` (but not really difficult with a triplestore storage)
- ?

**Comment by srosset81** [comment]
Any feedback @csarven @lecoqlibre ? Maybe overlooked advantages of hierarchical URIs ?

**Comment by csarven** [comment]
I've expressed some of these things elsewhere (that I'd have to dig up) but I'll try to respond to some things here that may be of help for what it is worth.

>But is it important for users to understand the inner workings of how data is structured by looking at the URI

That's not the use case as to why hierarchical paths are used but it is a useful outcome of the design. All things being equal, it is something that at least some users have some familiarity with because it is essentially what they also see in their operating systems - most common ones over decades follow the same design pattern. If anything, any URI is using slashes but without any imposed hierarchy is counter to what they are familiar with to the point that it may work as an anti-pattern. If some hierarchy is not desired in the URI, then there is virtually no reason to use them. All URIs following a URI Template like  `/{uuid}` would suffice, and it doesn't limit any system to have a "graph" view of things.

I would also add that there is a difference between resource organisation and knowledge organisation, and they don't need to be conflated. If some piece of information is to be categorised in a certain way, that's about the underlying knowledge, and not about the identifier.

> is it important for interoperating applications to manage that data easily

Yes, to some extent. The slash semantics is not so much about because it is for a filesystem but rather it is super useful for developers to re-use widely available existing libraries / tooling that understand 1) the hierarchy 2) can communicate with their system. This in essence enables "interoperability" on different levels (between classes of products and the environments they are in), and not just classes of products interoping among each other as they defined in a specification.

Just my take on things..

**Comment by csarven** [comment]
>    We can move the resource to another container without leaving behind dozens of redirects

Anything besides having all resources on the same container will be subjected to that "move" issue in a long enough timeline. So, either this is about knowledge organisation in which case information should be declarative part of the data or metadata, or it is about what the URI should be. When a URI is "allocated" to some thing (as per Web architecture), it carries a specific meaning. Redirecting it doesn't change the meaning. So, if something needs to be moved, it is because of other reasons.

>    If the container is renamed, we don't need to rename (eg. put redirects) its hundreds of contained resources

Why would the container name in the URI be "renamed"? The identifier is allocated. Why would that term need not be changed if not because of some knowledge reasons.

>    We can put the resources in several containers, and thus benefit from the various containers' default WAC permissions

Putting in different containers is about knowledge. Something can be discovered from different containers indeed, but that is orthogonal to how a particular resource is identified at a particular URI, and what that entails. I think different permissions potentially acting on the same resource will create more problems than provide an actual benefit but curious to see implementation feedback on that.

>    The ldp:contains predicate is the only source of truth about the resource containement

Right. With the way Solid Protocol uses LDP BC at least.

>    It's harder to guess what a resource is about (increased privacy)

If privacy is the driven factor of allocating a resource to a URI, then the terms in the path should not reveal anything pertaining to what the resource may be about. It is a marginal difference between using `/{uuid}` vs. `/foo/bar/baz` - if one doesn't want to supposedly disclose what the resource is about by looking at the URI. (I'm not suggesting that either is a better URI design.. just that there are different ways of looking at it all things equal.)

>    We are closer to blockchain-style identifiers, where everything is a hash (which will make sense if/when we use [NextGraph](https://nextgraph.org/) as a triple store)

Ok. Not sure why this is relevant unless there some kind of interop? Anything short of using an exact URI Template published by different systems would be non-interoperable by definition. So, there is no particular benefit even if URIs look like random looking gibberish.

>    You cannot have orphan resources, as you cannot detach a resource from a container (?)

Something is always under a container, for instance the root container.

>    It's easier for developers to know what the resource is about

Why is this relevant for developers? Developers shouldn't have to know anything about the resource by observing the URI.

>    It's easier to find the container of the resource, without needing to look for ldp:contains (Note: not really a problem with a triplestore storage !)

Right, the whole `..` idea that's virtually everywhere under the sun.

**Comment by nikoPLP** [comment]
Hello,

Niko here, from NextGraph (and a contributor to SemApps too, the base/ancestor of ActivityPods).

I find this topic of particular interest, and I am glad the discussion is happening.

I agree with the dichotomy between knowledge and resource organization established here by @csarven , it helps to clarify the reasoning.

We have clearly understood that Solid carries an opinionated view on resource allocation and identification, that is inherited from the "good old web" of 1991. We know very well where this is coming from, as the founder of Solid is no one else than the creator of the Web. A Web that was originally about creating static and hierarchical webpages on the disk of a webserver, and using that filesystem path, together with the server's address/domain, as a unique identifier for such resource. Moving and redirecting (301 and 302 HTTP codes) was allowed but cumbersome and not recommended (and didn't really pass the test of time, as the situation 35 years later is that dead links are ubiquitous).

But what worked well in the 90' and the early 2000, cannot be applied as a design pattern for all about data on the internet today. We will all agree I guess, that the vast majority of data shared on the internet today is not based on some hierarchical / "file-system like" organization anymore, thanks to the advent of scripting on the back-end and front-end. Modern applications exchange data with APIs that are non-hierarchical, that interact with databases and that are using HTTP just as a convenience (because that's what the browser can do) and switch to websocket as soon as they can. The identifiers of the resources being nowadays totally decoupled from the technical means used for their transport (which is what HTTP is essentially, and eventually) and their storage (the database, and not a static file anymore).

So to the previous dichotomy that we had so far between knowledge and resource, we could add transport, and I would argue that HTTP was originally binding resource identification and transport together (and possibly including some knowledge in the path and domain), and this strong correlation of the 3 concepts has long since gone.

I even find it a bit anachronistic that in 2024 we find ourselves still discussing if data should be identified in a hierarchical manner. End-users, that are maybe used to file-systems indeed, will never be bothered if the underlying systems they are using are not hierarchical. About the presentation of the data on some URL that appear in the address of the browser, we all know (and SEO people know it too) that the URL is just a representation that can change and will change, and can be redundant (many URLs linking to the same page or data) and so on. Everybody knows today that the URL is not a unique identifier of a resource.

Everything is now decoupled: knowledge, resource identification, transport, storage and representation.

The question we have to ask ourselves is about the Semantic Web principles, as Solid is an attempt at bringing the concepts of the Linked Data to the Web, its users, and to app developers. (And not the other way around! Solid should not try to bring the concepts of the (old) Web, to Linked Data, in my opinion).

Semantic Web advocates for a strong decoupling between data representation and its format, and allows knowledge to be encoded regardless of its current locator or place of birth. Anybody can establish facts about a resource in RDF, no matter if they are the original creator of such resource, or if their dataset resides in the same triplestore than the original one. It enables cross-server linking, much like hypertext was invented for, but the analogy stops here. Semantic Web is purely non hierarchical, and its resource identifiers are based, for now, on URLs, because this was so far the only available option. But interestingly, the specification of RDF does not mandate URLs, and any triple store in the world and any RDF library, can handle resource identifiers that are in fact URIs, and not URLs, and that leads me to where I want to go, which is that a URI of the form `urn:xxx` or `did:xxx` works perfectly fine in the Semantic Web, and we should rejoice about that, as it proves this technology was well designed and was able to pass the test of time.

The beauty of Semantic Web lies in the fact that the format of the identifier doesn't really matter, as all the data and metadata is expressed elsewhere, in triples that use the common identifier as an anchor, but do not put any meaning to it. The knowledge is expressed in the triples, not in the identifier. The representation is left out of scope of the specification. And the transport is so far, delegated to HTTP, with the dereferencing mechanism, but that could easily change too, when other types of identifiers are used (did, by example). 

A good example about that is the case of Wikidata, that did not encode any meaning/knowledge in the unique identifiers they use for resources. Instead, they use numerical indexes (Q80, by example, for Sir Tim Berners-Lee). From numeric indexes, we can easy pass to random UUID, conceptually, and practically too, as I demonstrate with NextGraph.
In the same manner, a resource can have one or several types (Classes) attached to it. 
What appears as a nightmare to taxonomists obsessed with building trees and categories, should not rebuke the developer who cares about data, meaning and knowledge. The Semantic Web is ready for such scenario when a resource can be understood as belonging to two or more classes or types concomitantly. Polymorphism being a concept programmers are familiar with, I am sure they can understand too that a resource can have many types. And I will come back to this a bit later, when addressing the question about LDP containers.

In the case of NextGraph, Random IDs, or more precisely cryptographic IDs, are needed for the underlying technology to operate. Using those IDs does extend the range of interoperability, and eventually decouples totally the identifier from any central authority (single point of failure, censorship, control). It is needed for local-first, end-to-end encryption, and adhoc features.

Coming back to the case of Solid, I find the efforts of advocating for Linked Data towards the advent of the Giant Global Graph of data envisioned by Tim Berners-Lee (aka Q80), noteworthy.

But we shouldn't try to fit square pegs into round holes, as Sebastian puts it.

The same goes, according to me, about the forgoing of SPARQL in the Solid specs. SPARQL is an integral part of the Semantic Web, that enables traversing the whole Global Graph at least within a single dataset, but even across the server's boundaries thanks to Federated Queries.

Why would then Solid confine the RDF and triples to static files that cannot be queries across files? (unless aggregated on the client side. but we shouldn't have to download the whole graph locally before being able to query it.)

This looks to me again as an attempt at bringing the concepts of the early static web into the semantic web. And not about bringing semantic concepts into the web, which would be much more desirable according to me.

We also have to understand that the web of today has nothing to compare with the one of the 90'.
We have seen the emergence of GraphQL, No-SQL, all kind of new models for data structures and novel databases. We know that "big-tech" is extensively using graph databases (and yes, I am saying databases, and not files on a disk, because querying the graph is important).

If we really want to address the topic of bringing the advent of Open formats and protocols for the Social Web and for the Giant Global Graph of data, we have to get rid of some old concepts that almost nobody is using anymore nowadays.

Knowledge and pure data are perfectly encoded with RDF, that also brings perfect interoperability and portability (once we get rid of URLs as resource identifiers).
Resource identifiers should be decentralized. DID is doing that very well.
Transport should be agnostic. HTTP is ok, more modern ways exist and should be available and used too.
Representation is up to everyone's taste and usage and will change over time, trend, language, location, technology, etc...

To address the specific question of containers in LDP, if I understood well, at least for LDP Basic Containers, one container is supposed to contain all the resources that have the same type.
At least if not in the specs, this is what is done in SemApps/ActivityPods.
RDF resources can have multiple <rdf:type>. So I don't see how we can enforce the rule of giving to that resource a unique identifier made of the URL of the server and the path of the container understood as the birthplace of such resource. Maybe the resource was born of type A, but will then change to become of type A+B, and then change again to be of type C. Shall we move it to other containers every time? Can it belong to two containers concurrently?

As you said, users are familiar with the concept of file-trees. This is also why we understand that the end-user might want to use the feature of LDP containers to organize their resources in folders and sub-folders (why not, after all). Likewise, an application developer might want to use Shape trees to store the data their application manipulates. All those concepts are gray areas for us (at least for me) as I am not sure of the added benefits of trees, over storing everything as flat triples in a quad-store. Anyway. Even in the case when hierarchical organization can be desirable. Wouldn't it be interesting to be able to "symlink" (I am using the filesystem analogy here on purpose) the resources from one folder into another and into many others, without any disruption?

If that was the case, then the unique identifier of the resource would not be a URL with a static path, but it would be a unique ID that the underlying system can understand, and that the end-user doesn't have to see. The analogy here with the filesystem is the concept of Inode, and in the solution that I propose, it is a DID unique cryptographic identifier, that can then be used to link the resource into folders, taxonomies, but also into endless connections and relationships between resources in order to form the Q3115452 (aka "Giant Global Graph").

I personally find that Solid has a great potential, and that bringing Semantic Web and Linked Data concepts to the world of programming and app development, in order to foment Open social networks, is a worthy goal.

But some self-imposed limitations that are currently present in the SOLID specs, are in fact akin to shooting oneself in the feet, and remove most of the benefits of using RDF.

I am available for further discussion on those important topics, here and on any other medium, if there is a will on the part of the Solid community to "free the specs" from such unfortunate chains.

**Comment by nikoPLP** [comment]
Hello,

Niko here, from NextGraph (and a contributor to SemApps too, the base/ancestor of ActivityPods).

I find this topic of particular interest, and I am glad the discussion is happening.

I agree with the dichotomy between knowledge and resource organization established here by @csarven , it helps to clarify the reasoning.

We have clearly understood that Solid carries an opinionated view on resource creation and identification, that is inherited from the "good old web" of 1991. We know very well where this is coming from, as the founder of Solid is no one else than the creator of the Web. A Web that was originally about creating static and hierarchical webpages on the disk of a webserver, and using that filesystem path, together with the server's address/domain, as a unique identifier for such resource. Moving and redirecting (301 and 302 HTTP codes) was allowed but cumbersome and not recommended (and didn't really pass the test of time, as the situation 35 years later is that dead links are ubiquitous).

But what worked well in the 90' and the early 2000, cannot be applied as a design pattern for all about data on the internet today. 

We will all agree I guess, that the vast majority of data shared on the internet today is not based on some hierarchical / "file-system like" organization anymore, thanks to the advent of scripting on the back-end and front-end. Modern applications exchange data with APIs that are non-hierarchical, that interact with databases and that are using HTTP just as a convenience (because that's what the browser can do) and switch to websocket as soon as they can. The identifiers of the resources being nowadays totally decoupled from the technical means used for their transport (which is what HTTP is essentially, and eventually: a transport protocol) and their storage (the database, and not a static file anymore).

So to the previous dichotomy that we had so far between knowledge and resource, we could add transport, and I would argue that HTTP was originally binding resource identification and transport together (and possibly including some knowledge in the path and domain), and this strong correlation of the 3 concepts has long since gone.

I even find it a bit anachronistic that in 2024 we find ourselves still discussing if data should be identified in a hierarchical manner. End-users, that are maybe used to file-systems indeed, will never be bothered if the underlying systems they are using are not hierarchical. About the presentation of the data on some URL that appears in the address bar of the browser, we all know (and SEO people know it too) that the URL is just a representation that can change and will change, and can be redundant (many URLs linking to the same page or data) and so on. Everybody knows today that the URL is not a unique identifier of a resource.

Everything is now decoupled: knowledge, resource identification, transport, storage and representation.

The question we have to ask ourselves is about the Semantic Web principles, as Solid is an attempt at bringing the concepts of Semantic/Linked Data to the Web, its users, and to app developers. (And not the other way around! Solid should not try to bring the concepts of the (old) Web, to Linked Data, in my opinion).

Semantic Web advocates for a strong decoupling between data representation and its format, and allows knowledge to be encoded regardless of its current locator or place of birth. Anybody can establish facts about a resource in RDF, no matter if they are the original creator of such resource, or if their dataset resides in the same triplestore than the original one. It enables cross-server linking, much like hypertext was invented for, but the analogy stops here. Semantic Web is purely non hierarchical, and its resource identifiers are based, for now, on URLs, because this was so far the only practical option. But interestingly, the specification of RDF does not mandate URLs, and any triple store in the world and any RDF library, can handle resource identifiers that are in fact URIs, and not URLs, and that leads me to where I want to go, which is that a URI of the form `urn:xxx` or `did:xxx` works perfectly fine in the Semantic Web, and we should rejoice about that, as it proves this technology was well designed and was able to pass the test of time.

The beauty of Semantic Web lies in the fact that the format of the identifier doesn't really matter, as all the data and metadata is expressed elsewhere, in triples that use the common identifier as an anchor, but do not put any meaning to it. The knowledge is expressed in the triples, not in the identifier. The representation is left out of scope of the specification. And the transport is so far, delegated to HTTP, with the dereferencing mechanism, but that could easily change too, when other types of identifiers are used (DID, by example). 

A good example about that is the case of Wikidata, that did not encode any meaning/knowledge in the unique identifiers they use for resources. Instead, they use numerical indexes (Q80, by example, for Sir Tim Berners-Lee). From numeric indexes, we can easily pass to random UUID, conceptually, and practically too, as I demonstrate with NextGraph.
In the same manner, a resource can have one or several types (Classes) attached to it. 
What appears as a nightmare to taxonomists obsessed with building trees and categories, should not rebuke the developer who cares about data, meaning and knowledge. The Semantic Web is ready for such scenario when a resource can be understood as belonging to two or more classes or types concomitantly. Polymorphism being a concept programmers are familiar with, I am sure they can understand too that a resource can have many types. And I will come back to this a bit later, when addressing the question about LDP containers.

In the case of NextGraph, Random IDs, or more precisely cryptographic IDs, are needed for the underlying technology to operate. Using those IDs does extend the range of interoperability, and eventually decouples totally the identifier from its creator and birth-place, and from any central authority (single point of failure, censorship, control). It is needed for local-first, end-to-end encryption, and adhoc features.

Coming back to the case of Solid, I find the efforts of advocating for Linked Data towards the advent of the Giant Global Graph of data envisioned by Tim Berners-Lee (aka Q80), noteworthy.

But we shouldn't try to fit square pegs into round holes, as Sebastian puts it.

The same goes, according to me, about the forgoing of SPARQL in the Solid specs. SPARQL is an integral part of the Semantic Web, that enables traversing the whole Global Graph at least within a single dataset, but even across the server's boundaries thanks to Federated Queries.

Why would then Solid confine the RDF and triples to static files that cannot be queried across files? (unless aggregated on the client side. but we shouldn't have to download the whole graph locally before being able to query it.)

This looks to me again as an attempt at bringing the concepts of the early static web into the semantic web. And not about bringing semantic concepts into the web, which would be much more desirable according to me.

We also have to understand that the web of today has nothing to compare with the one of the 90'.
We have seen the emergence of GraphQL, No-SQL, all kind of new models for data structures and novel databases. We know that "big-tech" is extensively using graph databases (and yes, I am saying databases, and not files on a disk, because querying the graph is important).

If we really want to address the topic of bringing the advent of Open formats and protocols for the Social Web and for the Giant Global Graph of data, we have to get rid of some old concepts that almost nobody is using anymore nowadays.

Knowledge and pure data are perfectly encoded with RDF, that also brings perfect interoperability and portability (once we get rid of URLs as resource identifiers).
Resource identifiers should be decentralized. DID is doing that very well.
Transport should be agnostic. HTTP is ok, more modern ways exist and should be available and used too.
Representation is up to everyone's taste and usage and will change over time, trend, language, location, technology, etc...

To address the specific question of containers in LDP, if I understood well, at least for LDP Basic Containers, one container is just a folder on the filesystem of the Solid server, and app developers can put data wherever they want in the PODS of the user. Then Shape Trees help to define which folder is used for what, and can attach a specific Class to a folder, so that other apps know where to find or modify the resources belonging to this Class.
RDF resources can have multiple classes by using the <rdf:type> predicate several times. So I don't see how we can enforce the rule of giving to that resource a unique identifier made of the URL of the server and the path of the container understood as the birthplace of such resource. Maybe the resource was born of type A, but will then change to become of type A+B, and then change again to be of type C. Shall we move it to other containers every time? Can it belong to two containers concurrently? Why should its unique identifier (URL) be tied to the container/folder it was originally created in, or to the class it used to have, while the type of the resource might have changed, and the container or folder that saw its birth might have been deleted or renamed since then?

As you said, users are familiar with the concept of file-trees. This is also why we understand that the end-user might want to use the feature of LDP containers to organize their resources in folders and sub-folders (why not, after all). Likewise, an application developer might want to use Shape trees to store the data their application manipulates. All those concepts are gray areas for us (at least for me) as I am not sure of the added benefits of trees, over storing everything as flat triples in a quad-store. Anyway. Even in the case when hierarchical organization can be desirable. Wouldn't it be interesting to be able to "symlink" (I am using the filesystem analogy here on purpose) the resources from one folder into another and into many others, without any disruption?

If that was the case, then the unique identifier of the resource would not be a URL with a static path, but it would be a unique ID that the underlying system can understand, and that the end-user doesn't have to see. The analogy here with the filesystem is the concept of Inode, and in the solution that I propose, it is a DID unique cryptographic identifier, that can then be used to link the resource into containers, folders, taxonomies, but also into endless connections and relationships between resources in order to form the Q3115452 (aka "Giant Global Graph").

I personally find that Solid has a great potential, and that bringing Semantic Web and Linked Data concepts to the world of programming and app development, in order to foment Open social networks, is a worthy goal.

But some self-imposed limitations that are currently present in the SOLID specs, are in fact akin to shooting oneself in the feet, and remove most of the benefits of using RDF.

I am available for further discussion on those important topics, here and on any other medium, if there is a will on the part of the Solid community to "free the specs" from such unfortunate chains.

**Comment by kjetilk** [comment]
Since one of my initial comments were referenced in the initial comment, I'll just share some of my evolved thinking.

Firstly, in a client-server architecture, the server wields much power, [excessive power, some might say](https://berjon.com/user-agency/). Thus, it is important that the protocol design does not constrain user's behavior unnecessarily.

In the current design, it is up to the user to design their URI space, if they want to use non-hierarchical URIs, they can do so by using UUIDs in the root container, for example. I think this is important. One shouldn't think that Tim didn't appreciate the importance of not getting constrained to hierarchies, it is the Web, after all. 

That things move around is inevitable, but it doesn't influence URIs if nothing in the URI has any semantics that change. This could point to UUIDs, but it also points in a different direction: Do not use the hierarchy for knowledge organization, use for example SKOS for that (now, Tim once told me that he likes using URIs for knowledge organization, to which I disagree). Presently, WAC attaches some utility to the hierarchy, and I have been concerned for a long time this would cause practical problems, but it is really the only thing in Solid that does that.

I think there's a coordination problem here: The user of the pod should have the authority to design their URI space. If the server only allows non-hierarchical URIs, then it will infringe on the user's right here, so I am against that. But if apps simply grabs some of the user's space without them having any influence, that too would infringe on that right. Ultimately, I suppose you could say that the user wouldn't care, but then, I am also concerned that Solid would become just a datastore where the user would have little power to ensure interop, and so interop wouldn't happen.

I would also remark that I agree that it is very unfortunate that SPARQL hasn't played a greater role. The main reason I joined was to have federated query over decentralized pods. I think that's crucial, but there was no interest in it. 

In conclusion I'm now :-1: on getting rid of hierarchical paths, that would be a constraint on the user's freedom. Instead, other technologies should be used on the top of the resources to add semantics to them.

**Comment by nikoPLP** [comment]
Nice to read your take on the matter @kjetilk ! And thank you for taking the time to answer.

Also I am glad to see that we share some common views about SPARQL and your concern for separation between knowledge/semantic and URIs. I know WAC quite well for having implemented some of the specs in SemApps, but i would say it is a minor issue that the paths there hold some semantic.

About the question of freedom for the end-user, if it is important that they can build paths for their URIs, then we could implement that in a virtual way for them. The URL with path would point internally (without the user seing that) to a UUID. The fact that we want to use UUID is not to impose anything to the user. It is just that internally, that's what we have. But if I take the analogy with Inodes again, then the Inode is hidden to the end-users, that only sees files and directories. So we could probably do the same here: we would emulate a hierarchy that can be used in the paths of URLs, and internally it would all be UUIDs.

One question that I still have is the understanding on what is supposed to happen if a URL contains some path (which I guess reflect a tree of containes. but maybe I am wrong here), then, if the paths are constructed with a hierarchy of containers, and the URL is full of those paths, and if we consider that the URL is the unique identifier to the resource and that it should be permanent and never change, then, what happens when the paths (containers) change after the resource was created and was assigned a URL? I find the whole correlation between URL and path a bit confusing and unnecessarily risky for long term consistency of the links.

Could you bring some light on that question that I have? How can we make the URLs durable, if they contain paths, and those paths are subject to arbitrary changes. Is the spec saying anything about that? Is it just left to the user to organize his pod as he pleases, and if he is renaming the paths/folders, then it is his fault and should expect some 404 ?

Can't we do a little bit better on that, to help the user not break his links?

**Comment by kjetilk** [comment]
> One question that I still have is the understanding on what is supposed to happen if a URL contains some path (which I guess reflect a tree of containers. but maybe I am wrong here), then, if the paths are constructed with a hierarchy of containers, and the URL is full of those paths, and if we consider that the URL is the unique identifier to the resource and that it should be permanent and never change, then, what happens when the paths (containers) change after the resource was created and was assigned a URL?

I think there are two answers to this, one is that you use a conventional `301` redirect. Make sure that it is automated to make that redirect, so that if it does happen, things don't break.

The other answer, I feel, is that you don't attach too much meaning to the URI. Nothing that the user sees should rely on the URI (which, I suppose, is the case with UUIDs), but you may use internal structure of the URI for management within apps, that's fine as long as you don't rely on it being static. But best practice should be to rely on the RDF describing what the resource semantics and use that.

But then, I stopped developing on Solid, so it is ultimately implementation experience that counts.
